metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala

```diff
@@ -1087,10 +1106,60 @@ class MetalsLanguageServer(
             )
           )
         }.asJavaObject
+      case ServerCommands.StartDebugAdapter() =>
+        val args = params.getArguments.asScala
+        args match {
+          case Seq(param) =>
+            param.as[b.DebugSessionParams] match {
+              case Failure(exception) =>
+                Future.failed(exception).asJavaObject
+              case Success(parameters) =>
+                val proxyServer = new ServerSocket(0)
+
+                val awaitClient = () => {
+                  Future(proxyServer.accept()).withTimeout(10, TimeUnit.SECONDS)
+                }
+
+                val connectToServer = () => {
+                  val debugSession = buildServer
+                    .map(_.startDebugSession(parameters).asScala)
+                    .getOrElse(
+                      Future
+                        .failed(new IllegalStateException("No build server"))
+                    )
+
+                  debugSession
+                    .withTimeout(10, TimeUnit.SECONDS)
+                    .map(uri => {
+                      val socket = new Socket()
+
+                      val address =
+                        new InetSocketAddress(uri.getHost, uri.getPort)
+                      val timeout = TimeUnit.SECONDS.toMillis(10).toInt
+                      socket.connect(address, timeout)
+
+                      socket
+                    })
+                }
+
+                val server = DebugServer.create(awaitClient, connectToServer)
```

Comment:
From what I see both `awaitClient` and `connectToServer` both timeout - could we just merge the futures and await for them in DebugServer?